evm/src/NttManager/NttManagerWithPerChainTransceivers.sol

```diff
@@ -0,0 +1,165 @@
+    }
+
+    function _getPerChainTransceiverBitmap(
+        uint16 forChainId,
+        bytes32 tag
+    ) private view returns (uint64 bitmap) {
+        bitmap = _getPerChainTransceiverBitmapStorage(tag)[forChainId].bitmap;
+        if (bitmap == 0) {
+            bitmap = type(uint64).max;
+        }
+    }
+
+    function _getPerChainTransceiverBitmapStorage(
+        bytes32 tag
+    ) private pure returns (mapping(uint16 => _EnabledTransceiverBitmap) storage $) {
+        uint256 slot = uint256(tag);
+        assembly ("memory-safe") {
+            $.slot := slot
+        }
+    }
+
+    // ==================== Override / implementation of threshold stuff =========================
+
+    /// @inheritdoc IManagerBase
+    function setPerChainThreshold(
+        uint16 forChainId,
+        uint8 threshold
+    ) external override(ManagerBase, IManagerBase) onlyOwner {
+        if (threshold == 0) {
+            revert ZeroThreshold();
+        }
+
+        mapping(uint16 => _Threshold) storage _threshold = _getThresholdStoragePerChain();
+        uint8 oldThreshold = _threshold[forChainId].num;
+
+        _threshold[forChainId].num = threshold;
+        _checkThresholdInvariants(_threshold[forChainId].num);
+
+        emit PerChainThresholdChanged(forChainId, oldThreshold, threshold);
+    }
+
+    function getPerChainThreshold(
+        uint16 forChainId
+    ) public view override(ManagerBase, IManagerBase) returns (uint8) {
+        uint8 threshold = _getThresholdStoragePerChain()[forChainId].num;
+        if (threshold == 0) {
+            return _getThresholdStorage().num;
+        }
+        return threshold;
+    }
+
+    function _getThresholdStoragePerChain()
+        private
+        pure
+        returns (mapping(uint16 => _Threshold) storage $)
+    {
+        // TODO: this is safe (reusing the storage slot, because the mapping
+        // doesn't write into the slot itself) buy maybe we shouldn't?
+        uint256 slot = uint256(THRESHOLD_SLOT);
+        assembly ("memory-safe") {
+            $.slot := slot
+        }
+    }
+
+    /// @inheritdoc IManagerBase
+    function isMessageApproved(
+        bytes32 digest
+    ) public view override(ManagerBase, IManagerBase) returns (bool) {
+        uint16 sourceChainId = _getMessageAttestationsStorage()[digest].sourceChainId;
+        uint8 threshold = getPerChainThreshold(sourceChainId);
+        return messageAttestations(digest) >= threshold && threshold > 0;
+    }
+
+    function _getMessageAttestations(
+        bytes32 digest
+    ) internal view override returns (uint64) {
+        AttestationInfo memory attInfo = _getMessageAttestationsStorage()[digest];
+        uint64 enabledTransceiverBitmap = _getEnabledTransceiversBitmap();
+        uint64 enabledTransceiversForChain =
+            _getEnabledRecvTransceiversForChain(attInfo.sourceChainId);
+        return attInfo.attestedTransceivers & enabledTransceiverBitmap & enabledTransceiversForChain;
```

Comment:
Nice, at a glance I think this may guard against my previous comment. Is defaulting the bitmap to `type(uint64).max` above instead of loading the enabled transceivers deliberate to save gas?

Reply:
Yeah, returning max does avoid a storage read, and makes this work. I can either add a comment on `_getPerChainTransceiverBitmap` to explain that or add the extra storage read in `_getPerChainTransceiverBitmap`. How strongly do you feel about it?

Reply:
I'm ok with adding some inline docs to explain it, think it's a neat way to save some gas